plugin_system/plugins/gregor_plugin.py

```diff
@@ -0,0 +1,144 @@
+    """
+    Plugin for GREGoR U08 release data on Terra
+    """
+
+    def create_sample_phenotype_index(
+        self, phenotype_table: str = None, as_set: bool = True, save_path: str = None
+    ) -> dict[str, list[str] | set[str]]:
+        """given phenotypical data input specified by the GREGoR Data model (in either tsv/csv or Terra data table),
+        return a dict mapping from each sample id to the sample's list of phenotypes
+
+        Returns:
+            dict[str, list[str]]: index of a sample id to sample's phenotypes
+        """
+
+        if phenotype_table is None:
+            # if unspecified, ensure valid Terra environment
+            for env_key in ["WORKSPACE_NAMESPACE", "WORKSPACE_NAME"]:
+                assert (
+                    env_key in os.environ
+                ), f"ERROR: No {env_key} key found in environmental variables in the Terra workspace. If you are working in a Terra workspace, please ensure both a WORKSPACE_NAMESPACE and a WORKSPACE_NAME are specified."
+
+            # create dataframe from Terra data table
+            # https://github.com/broadinstitute/fiss/blob/master/firecloud/api.py
+            try:
+                response = fapi.get_entities_tsv(
+                    os.environ["WORKSPACE_NAMESPACE"],
+                    os.environ["WORKSPACE_NAME"],
+                    "phenotype",
+                    model="flexible",
+                )
+                response.raise_for_status()
+            except Exception as e:
+                if response.json() and e in response.json():
+                    error_message = response.json()["message"]
+                else:
+                    error_message = e
+                print(
+                    f"Error while loading phenotype data table from workspace: \n{error_message}"
+                )
+
+            phenotype_tsv = io.StringIO(response.text)
+            phenotype_df = pd.read_csv(phenotype_tsv, sep="\t")
+        else:
+            # table path specified, parse using that table
+            with open(phenotype_table, "r") as file:
+                if phenotype_table.endswith(".csv"):
+                    phenotype_df = pd.read_csv(file)
+                elif phenotype_table.endswith(".tsv"):
+                    phenotype_df = pd.read_csv(file, sep="\t")
+                else:
+                    raise Exception(
+                        "Only csv and tsv file types implemented for phenotype table"
+                    )
+
+        # create patient to phenotype dict
+        phenotype_index = {}
+        for participant_id in phenotype_df["participant_id"].unique():
+            phenotypes = phenotype_df[phenotype_df["participant_id"] == participant_id][
+                "term_id"
+            ].unique()
+
+            phenotype_index[participant_id] = (
+                set(phenotypes) if as_set else list(phenotypes)
+            )
+
+        # save to disk if specified
+        if save_path:
```

Comment:
what's the utility of this? If the idea is to be able to reload a file from local disk to make repeated queries cheaper, is there a way to reload that local copy?

Reply:
I think this is some messiness that came from previous functionality, I might make it into a util and better incorporate it into the GREGoR plugin!

Reply:
fixed and moved to utils